pkg/kudoctl/kudoinit/setup/setup.go

```diff
@@ -39,6 +42,79 @@ func NewInstaller(options kudoinit.Options, crdOnly bool) *Installer {
 	}
 }
 
+// Validate checks that the current KUDO installation is correct
+func (i *Installer) VerifyInstallation(client *kube.Client, result *verifier.Result) error {
+	// Check if all steps are correctly installed
+	for _, initStep := range i.steps {
+		if err := initStep.VerifyInstallation(client, result); err != nil {
+			return fmt.Errorf("error while verifying init step %s: %v", initStep.String(), err)
+		}
+	}
+
+	return nil
+}
+
+func requiredMigrations() []migration.Migrater {
+
+	// TODO: Determine which migrations to run
+	return []migration.Migrater{
+		// TODO: Implement actual migrations
+	}
+}
+
+func (i *Installer) PreUpgradeVerify(client *kube.Client, result *verifier.Result) error {
+	// Step 1 - Verify that upgrade can be done
+	// Check if all steps are upgradeable
+	for _, initStep := range i.steps {
+		if err := initStep.PreUpgradeVerify(client, result); err != nil {
+			return fmt.Errorf("error while verifying upgrade step %s: %v", initStep.String(), err)
+		}
+	}
+	if !result.IsValid() {
+		return nil
+	}
+
+	// Step 2 - Verify that any migration is possible
+	migrations := requiredMigrations()
+	clog.Printf("Verify that %d required migrations can be applied", len(migrations))
+	for _, m := range migrations {
+		if err := m.CanMigrate(client); err != nil {
```

Comment:
I think, there is more to migrations:
- can we only upgrade or also downgrade?
- do we need to check the versions?
- how do you know which migrations are required e.g. when a migration skips a few versions?

I don't think we need to tackle all that right now. This PR already adds value in the form of manager upgrades. I would be ok, leaving out migrations for now and concentrating on upgrades and verification.

Reply:
Oh yes, there's a lot missing with regards to migrations, but I think that's better left for another PR